examples/benchmarks/other/unify.effekt

```diff
@@ -0,0 +1,178 @@
+        do raise(UnificationFailure(substTy1, substTy2), "Different constructors!")
+      } else if (args1.size != args2.size) {
+        do raise(UnificationFailure(substTy1, substTy2), "Different number of arguments!")
+      } else {
+        unifyMany(args1, args2, subst)
+      }
+  }
+}
+
+// Unify a list of arguments with a current substitution
+def unifyMany(args1: List[Type], args2: List[Type], subst: Substitution): Substitution / Exception[UnificationError] =
+  (args1, args2) match {
+    case (Nil(), Nil()) => subst
+    case (Cons(a1, rest1), Cons(a2, rest2)) =>
+      val newSubst = unify(a1, a2, subst)
+      unifyMany(rest1, rest2, newSubst)
+    case _ => do raise(UnificationManyFailure(args1, args2), "Different numbers of types on each side!")
+  }
+
+def unify(ty1: Type, ty2: Type): Substitution / Exception[UnificationError] =
+  unify(ty1, ty2, map::empty(box bytearray::compareStringBytes))
+
+def showType(ty: Type): String = ty match {
+  case Var(name) => name
+  case Con(name, Nil()) => name
+  case Con(name, args) =>
+    name ++ "(" ++ args.map { t => showType(t) }.join(", ") ++ ")"
+}
+
+def show(err: UnificationError): String = err match {
+  case OccursCheckFailure(variable, ty) =>
+    "Occurs check failed: " ++ variable ++ " occurs in " ++ showType(ty)
+  case UnificationFailure(ty1, ty2) =>
+    "Cannot unify " ++ showType(ty1) ++ " with " ++ showType(ty2)
+    case UnificationManyFailure(tps1, tps2) =>
+    "Cannot unify " ++ tps2.map { showType }.join(", ") ++ " with " ++ tps1.map { showType }.join(", ")
+}
+
+/// Worker wrapper
+def reporting { body : => Substitution / Exception[UnificationError] }: Unit / emit[(String, Type)] = {
+  val res = result[Substitution, UnificationError] {
+    body()
+  }
+
+  res match {
+    case Success(subst) => {
+      println("Unification successful!")
+      subst.each
+    }
+    case Error(err, msg) =>
+      println("Unification failed: " ++ show(err))
+      if (msg.length > 0) {
+        println(msg)
+      }
+  }
+}
+
+/// Used for testing to generate two `depth`-deep, nested types of the shape:
+/// ```
+/// (Nested
+///   (Nested
+///     ...
+///       XLLLLLLLL
+///       XLLLLLLLR)
+/// ```
+/// for `baseVar = X`.
+def generateDeepType(depth: Int, baseVar: String): Type = {
+  def recur(currentDepth: Int, varSuffix: String): Type =
+    if (currentDepth == 0) {
+      Var(baseVar ++ varSuffix)
+    } else {
+      Con("Nested", [
+        recur(currentDepth - 1, varSuffix ++ "L"),
+        recur(currentDepth - 1, varSuffix ++ "R")
+      ])
+    }
+
+  recur(depth, "")
+}
+
+def main() = {
+  def printBinding(pair: (String, Type)): Unit =
+    println("  " ++ pair.first ++ " -> " ++ showType(pair.second))
+
+  // sanity check
+  for {
+    reporting {
+      val intType = Con("Int", [])
+      val listType = Con("List", [intType])
+      val typeVar = Var("a")
+
+      unify(typeVar, listType)
+    }
+  } { printBinding }
+
+  // the actual test
+  val N = 12
```

Comment:
`N = 12` means unification of two trees of depth 12, i.e. tree of tuples with `2^12 = 4096` variables in the leaves.
We can tweak this number or even use the `benchmarks/runner`...